File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.py
def solution(friends, gifts):
    gift_s = {friend : [] for friend in friends}
    gift_r = {friend : 0 for friend in friends}
    for gift in gifts:
        s, r = gift.split()
        gift_s[s] = []

    for gift in gifts:
        s, r = gift.split()
        gift_r[s] += 1
        gift_r[r] -= 1
        gift_s[s].append(r)

    n = len(friends)
    graph = [[0]*(n) for _ in range(n)]
    hashed_friend = {friends[i]: i for i in range(n)} 
    
    # 그래프 만들기 
    for sender in gift_s:
        for reciever in gift_s[sender]:
            graph[hashed_friend[sender]][hashed_friend[reciever]] += 1 
    
    answer = 0
    gift = [0]*n
    for i in range(n):
        for j in range(n):
            if i != j:
                if graph[i][j] != graph[j][i]:
                    if graph[i][j] > graph[j][i]:
                        gift[i] += 1
                else: 
                    if gift_r[friends[i]] != gift_r[friends[j]]:
                        if gift_r[friends[i]] > gift_r[friends[j]]: 
                            gift[i] += 1
    return max(gift)

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.js
function solution(friends, gifts) {
    const n = friends.length;
    const gift_lst = {};
    const gift_zisu = {};
    friends.forEach((friend, i) => {
        gift_lst[friend] = i;
        gift_zisu[friend] = 0;
    });

    const graph = Array.from({ length: n }, () => Array(n).fill(0));

    gifts.forEach(gift => {
        const [send, receive] = gift.split(" ");
        graph[gift_lst[send]][gift_lst[receive]] += 1;
        gift_zisu[send] += 1;
        gift_zisu[receive] -= 1;
    });

    const answer = new Array(n).fill(0);
    for (let i = 0; i < n; i++) {
        for (let j = 0; j < n; j++) {
            if (i !== j) {
                if (graph[i][j] > graph[j][i]) {
                    answer[i] += 1;
                } else if (graph[i][j] === graph[j][i]) {
                    if (gift_zisu[friends[i]] > gift_zisu[friends[j]]) {
                        answer[i] += 1;
                    }
                }
            }
        }
    }

    return Math.max(...answer);
}
